File: 1.two-sum.go
/*
 * @lc app=leetcode id=1 lang=golang
 *
 * [1] Two Sum
 */

// @lc code=start

package main

func twoSum(nums []int, target int) []int {
	//key: difference
	//value: indexInMap
	var checkMap = make(map[int]int)

	//i: index
	for i, numInNums := range nums {
		//return when one of difference + num == target
		if indexInMap, ok := checkMap[numInNums]; ok {
			return []int{indexInMap, i}
		}
		//store difference and index
		checkMap[target-numInNums] = i
	}
	return nil
}

// @lc code=end

File: python-code/leetCode/1.twoSum.py
# https://leetcode.com/problems/two-sum/description/
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        num_map = dict()
        for idx, val in enumerate(nums):
            if val not in num_map:
                num_map[val] = []
            num_map[val].append(idx)
        for i, v in enumerate(nums):
            diff = target - v
            if diff not in num_map:
                continue
            for j in num_map[diff]:
                if i == j:
                    continue
                return [i, j]
        return []


solution = Solution()
print(solution.twoSum([2, 7, 11, 15], 9))  # 2
# nums = [2,7,11,15], target = 9
